rangeserver/src/wal/cassandra.rs

```diff
@@ -0,0 +1,343 @@
+use std::ops::Deref;
+use std::ops::DerefMut;
+use std::sync::Arc;
+
+use crate::for_testing::in_memory_wal::InMemIterator;
+
+use super::*;
+use async_trait::async_trait;
+use flatbuffers::FlatBufferBuilder;
+use scylla::batch::Batch;
+use scylla::query::Query;
+use scylla::statement::SerialConsistency;
+use scylla::transport::errors::DbError;
+use scylla::transport::errors::QueryError;
+use scylla::Session;
+use scylla::SessionBuilder;
+use tokio::sync::RwLock;
+use uuid::Uuid;
+
+pub struct Cassandra {
+    session: Session,
+    wal_id: Uuid,
+    state: RwLock<State>,
+}
+
+enum State {
+    NotSynced,
+    Synced(LogState),
+}
+
+struct LogState {
+    first_offset: Option<i64>,
+    next_offset: i64,
+    flatbuf_builder: FlatBufferBuilder<'static>,
+}
+
+fn scylla_query_error_to_wal_error(qe: QueryError) -> Error {
+    match qe {
+        QueryError::TimeoutError | QueryError::DbError(DbError::WriteTimeout { .. }, _) => {
+            Error::Timeout
+        }
+        _ => Error::InternalError(Arc::new(qe)),
+    }
+}
+
+static SYNC_WAL_QUERY: &str = r#"
+  SELECT first_offset, next_offset FROM chardonnay.wal
+    WHERE wal_id = ? and offset = ?;
```

Comment:
Does Cassandra support named parameters? Positional is too error-prone imo.

Reply:
Looks like scylla db's client does provide that, I'll do a separate refactor to use in all the cassandra implementations we have:
https://rust-driver.docs.scylladb.com/stable/queries/values.html